### ISSUE TITLE
non-issue: request for Linux support

### ISSUE THREAD

**Issue by meoso** [bw1129/PIDtoolbox#6]
i see Linux MATLAB run-time is available.  it would be a nice feature to have this project run native in Linux too.

**Comment by bw1129** [comment]
thanks meoso. the thing is I'd need a linux machine and a version of matlab running on it to compile a linus version (not just runtime). That said, I should be able to install linus on a Mac using parallels or something right? I do have access to Matlab for linus from the University, so I will try to get this! cheers!

**Comment by ghostface** [comment]
Yes I did some reading and you can't crosscompile you need to compile it from a linux matlab installation.
I am currently trying to setup matlab on my box using the trial version - would be awesome if we can get linux binaries tho.
Not sure if I can compile something with the trial, if I do I'll add the bins here.

**Comment by ghostface** [comment]
I got it running under linux at least.
Caveats:

- blackbox_decode.exe is not found, I changed it in PTimport.m to blackbox_decode
- blackbox log files need to be in the same folder as PIDtoolbox - they won't load from anywhere else
-  scaling was messed up for me tiny matlab window, and huge PIDtoolbox fonts (i am running a 4k display with i3 as window manager so might not be an issue on other wms. I ran this to fix it somewhat in the commandline:

`
s = settings;s.matlab.desktop.DisplayScaleFactor
s.matlab.desktop.DisplayScaleFactor.PersonalValue = 2
`

**Comment by bw1129** [comment]
You’ll need bb decode for Linux. Just go to bb tools in GitHub and pick that up. Then you change system function to run bb decode in linux.

Wrt font and other sizes, go to lines in UI control scripts that state something about proportion screen size. It’s currently .065 or something. Changing this will have a global effect on changing button and all ui widgets. I think it even affects fonts.

**Comment by bw1129** [comment]
You might also need some toolboxes but hopefully you have em. Anyway, great job! Love to see how it works

**Comment by watageek** [comment]
For linux,
Can't you just use a distro on a live boot usb. (linux mint for example with rufus.exe)
16gb usb is easy to find

**Comment by bw1129** [comment]
@watageek I think you still have to have matlab for Linux installed with all the toolboxes. I have matlab for Linux. Have only Mac and Windows machines. I know I can install Linux on a Mac partition somehow but have not done it before. If u can point me in the right direction I might be able to get this moving.

**Comment by watageek** [comment]
Info is to be verified, on top of my head... There is a "folder" called persistence when you do a usb key with rufus.exe on a windows OS. Adjust to suit your needs.

**Comment by ghostface** [comment]
> You’ll need bb decode for Linux. Just go to bb tools in GitHub and pick that up. Then you change system function to run bb decode in linux.
> 
> Wrt font and other sizes, go to lines in UI control scripts that state something about proportion screen size. It’s currently .065 or something. Changing this will have a global effect on changing button and all ui widgets. I think it even affects fonts.

Yeah I got running just fine - thx for the tip about the proportion size - will check that out.

**Comment by watageek** [comment]
I run linux mint on a laptop PC on win10.  Dual boot. Grub menu is set to boot on linux default.

One of my friend is providing free unlimited 24/7 support. I am blessed... 😂

**Comment by RustyToad** [comment]
@bw1129 Here is a good guide to dual-booting Mac and Ubuntu. I used to have a mac and this method worked well for me for years.

I only run Ubuntu linux now, but am happy to help test any way I can. I'd love to get this running.

**Comment by watageek** [comment]
Post the link to that guide. 😚
 @RustyToad

**Comment by RustyToad** [comment]
@watageek sorry, just saw your reply. I don't know if I forgot the link, or if it was filtered out. Here it is again:

https://www.maketecheasier.com/install-dual-boot-ubuntu-mac/

If that doesn't work, it's the first hit on Google for me for "dual boot Mac Ubuntu". It's on maketecheasier, and uses REFInd to manage the boot up.

**Comment by meoso** [comment]
@bw1129 ... hey man, i see people telling how to dual-boot etc -- what a PITA...  If you do decide to support linux, then just install some linux distro (Xubuntu for ease) as a VM. (i..e Parallels if you already own it, or VirtualBox is free.) This is arguably "better" because you won't be outside of your normal OS environment, with the caveat that it's slower to run the VM instance, but there are some benefits such as copy/past or even sharing a folder between host-OS and VM.

But I digress, linux support would be great, but certainly should not take your priority. I can just as easily run a Windblows VM too.

**Comment by bw1129** [comment]
@meoso ya, I would do it using ubuntu or something. either way Id need to extend the matlab license and install for that. that's the main issue. I think I have it free from my uni so may do that, but for now Im busy with basic upgrades. the other thing though is every operating system using matlab requires slightly different programming, so I cant just copy past the code, but have to tailor it for Win, Mac, and Linux separately. that's tedious and prone to errors. anyway. will do my best

**Comment by qratz** [comment]
> I got it running under linux at least.
> Caveats:
> 
> * blackbox_decode.exe is not found, I changed it in PTimport.m to blackbox_decode
> * blackbox log files need to be in the same folder as PIDtoolbox - they won't load from anywhere else
> * scaling was messed up for me tiny matlab window, and huge PIDtoolbox fonts (i am running a 4k display with i3 as window manager so might not be an issue on other wms. I ran this to fix it somewhat in the commandline:
> 
> `s = settings;s.matlab.desktop.DisplayScaleFactor s.matlab.desktop.DisplayScaleFactor.PersonalValue = 2`

So, could you compile and offer linux binaries that we can use with the linux version of matlab runtime?

**Comment by ghostface** [comment]
> > I got it running under linux at least.
> > Caveats:
> > 
> > * blackbox_decode.exe is not found, I changed it in PTimport.m to blackbox_decode
> > * blackbox log files need to be in the same folder as PIDtoolbox - they won't load from anywhere else
> > * scaling was messed up for me tiny matlab window, and huge PIDtoolbox fonts (i am running a 4k display with i3 as window manager so might not be an issue on other wms. I ran this to fix it somewhat in the commandline:
> > 
> > `s = settings;s.matlab.desktop.DisplayScaleFactor s.matlab.desktop.DisplayScaleFactor.PersonalValue = 2`
> 
> So, could you compile and offer linux binaries that we can use with the linux version of matlab runtime?

As far as I understood you need a proper license to do this sadly. The evaluation one doesn't allow you to pre-combile binaries.

**Comment by bw1129** [comment]
@ghostface yes, I can get Matlab for linux. Working on it

**Comment by qratz** [comment]
in the meantime, i am working on porting to octave... so far so good:
<img width="960" alt="pidtoolbox_linux" src="https://user-images.githubusercontent.com/3747823/60726499-fe4f2f80-9f3b-11e9-99f3-e270db5cf3e9.png">

**Comment by bw1129** [comment]
@qratz Dude that's awesome! I have a new version ready for windows. Would you like the source code to bring this up to speed on the latest? Great work!

**Comment by qratz** [comment]
So far I have gotten the main view and the pid error view to work. Next I will try the spectograph. Once I have that working I'll put it on GitHub and try to keep track with the newest Version. The work is kind of a PITA, because there are so many tiny little things that work in octave but not in matlab, and vice versa. If you want to make life easier for me you could exclusively use the "set" and "get" methods on any GUI object properties, such as "foreground" and "fontsize".  As far as I know matlab accepts either something like guihandles.checkbox0.Value or get(guihandles.checkbox0, 'value') but octave only accepts the latter. Same goes for assignments via "set" vs. "=". That is like 90% of the changes I had to apply 😀.

**Comment by bw1129** [comment]
@qratz I think the main reason I aborted octave was that it tended to be very slow plotting, but that might be resolved with newer versions. can you confirm whether it is faster? e.g. when clicking different traces it would think for 1-2 secs. I think it was an opengl issue with octave
https://savannah.gnu.org/bugs/?53845

**Comment by qratz** [comment]
I will have to check that. For me, being a Linux only user, it was about having the toolbox at all rather than concerns about speed 😀. So far, I have only ran it on my laptop , which is an older i7 with Intel graphics, so basically no hardware accelleration. However, the startup time and initial loading of the BBL, feels very fast (have to go by feel here, no benchmarks available yet). But you are right, when I click on one of the plots on the main view it takes a second or two to display that plot zoomed in.

**Comment by ghostface** [comment]
@qratz can you please share your progress on the octave port for linux?

**Comment by qratz** [comment]
Hi everyone,
i had to suspend my work on the octave port for a few weeks, because i had to move :). i have been working my way through the menu system of pidtoolbox, starting with the main page and then the piderror and step response, leaving the last and biggest chunk last: the spectograph. There is a bunch of stuff there that is unsupported by octave. One problem i have not really focused on is the way the font sizes are calculated. This seams to be overly complicated in the code and leads to font sizes adjusting whenever i reopen a tool window. This looks extremly bad in octave, but the same problem is there in the windows version of original pidtoolbox, but not to the same extent as in octave.

Brian, have you noticed that too?

**Comment by bw1129** [comment]
Yes, the font thing is unnecessary, but I tried to make it so that each time the screen size is changed, the fonts would change accordingly. This is especially useful for differences in visibility across different screen sizes, but it’s certainly not essential. My advice would be to fix a font size to something like 12 or 14 point an go with that.

**Comment by bw1129** [comment]
Yes, the font thing is unnecessary, but I tried to make it so that each time the screen size is changed, the fonts would change accordingly. This is especially useful for differences in visibility across different screen sizes, but it’s certainly not essential. My advice would be to fix a font size to something like 12 or 14 point an go with that.

**Comment by bw1129** [comment]
@tiagolramos  I wish! thanks for the offer. I just need to install Linux and Matlab for Linux with the appropriate toolboxes, that's all. aiming to do that soon!

**Comment by bw1129** [comment]
Would anyone here be willing to compile a version of blackbox_decode for Linux (Ubuntu)? This would really help. I’m hoping to compile Ptb in Linux very soon but would need the package blackbox_decode with it.

**Comment by tiagolramos** [comment]
compiled in ubuntu 18.04

[blackbox-tools-linux.tar.gz](https://github.com/bw1129/PIDtoolbox/files/3737155/blackbox-tools-linux.tar.gz)

**Comment by bw1129** [comment]
Oh man, thanks so much! Will try to compile Linux version of Ptb today

Sent from my iPhone

On Oct 16, 2019, at 10:36 PM, tiagolramos <notifications@github.com<mailto:notifications@github.com>> wrote:


compiled in ubuntu 18.04

blackbox-tools-linux.tar.gz<https://can01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fbw1129%2FPIDtoolbox%2Ffiles%2F3737155%2Fblackbox-tools-linux.tar.gz&data=02%7C01%7Cbrian.white%40queensu.ca%7C1de3f06a0e2a4df5d7a008d752aad677%7Cd61ecb3b38b142d582c4efb2838b925c%7C1%7C0%7C637068766012430831&sdata=oktUPCo9JpQJMr5Vbb8%2Fk%2FyX6WwHmlcQtuqqfZENlB8%3D&reserved=0>

**Comment by bw1129** [comment]
A BETA version of PIDtoolbox for Linux is available for download. Please follow the download instructions and let me know if it works. I was unable to test the executable directly (only tested from Matlab itself and it ran). 
https://github.com/bw1129/PIDtoolbox/releases/tag/v0.37

**Comment by meoso** [comment]
installed with your included installer.
had to `chmod +x` a couple things
installed with `sudo` to default folders (`/usr/PIDtoolbox` & `/usr/local/MATLAB/MATLAB_Runtime`)
tried `./run_PIDtoolbox.sh /usr/local/MATLAB/MATLAB_Runtime/v97` from both the extracted `./main` & `/usr/PIDtoolbox/application`
result:
```
./run_PIDtoolbox.sh /usr/local/MATLAB/MATLAB_Runtime/v97
------------------------------------------
Setting up environment variables
---
LD_LIBRARY_PATH is .:/usr/local/MATLAB/MATLAB_Runtime/v97/runtime/glnxa64:/usr/local/MATLAB/MATLAB_Runtime/v97/bin/glnxa64:/usr/local/MATLAB/MATLAB_Runtime/v97/sys/os/glnxa64:/usr/local/MATLAB/MATLAB_Runtime/v97/sys/opengl/lib/glnxa64
Error using matlabpath
Modifying the search path is not supported by MATLAB Compiler. Remove functions that modify the search path from your MATLAB code. To make files visible to your deployed application, add the parent folder to your MATLAB session.

Error in path (line 109)

Error in addpath (line 86)

Error in PIDtoolbox (line 14)

MATLAB:mpath:PathAlterationNotSupported

```

**Comment by tiagolramos** [comment]
i can try make a deb package after the installation script is fine tuned

**Comment by vaughnd** [comment]
Same trouble with the 'modifying the search path' for me that meoso has.

**Comment by mituritsyn** [comment]
@qratz hello, great work with octave port! Maybe will be better if someone could help you? Is your progress available on git or anywhere?
@bw1129 thanks for this progect. Is there any idea how to make linux reliases regular?

**Comment by qratz** [comment]
> @qratz hello, great work with octave port! Maybe will be better if someone could help you? Is your progress available on git or anywhere?
> @bw1129 thanks for this progect. Is there any idea how to make linux reliases regular?

Since folks started publishing Linux releases I stopped working on the octave port. But that was not the only reason. I kind of got a brick wall with some of the analytical tools used in Matlab which are not available in octave, especially in the spectograph. Due to work I did not have the time to dig deeper into octave. But since we are now all under stay at home order I might start to work on it again. The thing that cost so much time, was actually the thousands of tiny octave-matlab thing. It would be a HUGE help if the original code constantly used set and get instead of "." To access properties.

**Comment by bw1129** [comment]
@mituritsyn @qratz I think the Linux release had some issues and I couldnt find the time to resolve them. wrt octave, I actually had a running version, but what held me back was mostly the slow plotting. I may dig this up again and see if I can release what I have. I had the main stuff working (spectrograms and step plots). there were just aspects of it I wasnt happy with at the time. anyway, I'll dig into what I have and see if I can upload a decently working version. thanks for reminding me.

**Comment by mituritsyn** [comment]
@qratz @bw1129 as I understood, you guys both make a huge work on octave independently. Just imagine, what could happen in coop or even with community help. Get out of the sandbox)

**Comment by bw1129** [comment]
@MITuritsyn it’s not an issue with sharing. I simply didn’t want to release a bunch of messy code and this code is based on a very old version of Ptb so that would not be helpful either. I released Ptb so I have no issue with releasing an octave version. Just waiting till it’s functional enough to be useable, and time is precious these days.

**Comment by flexd** [comment]
What is missing to make this work? Is there anything I can help with?

**Comment by ghostface** [comment]
The problem is this:

```
executableDir = pwd;
addpath(executableDir)
```

It causes the compiled binary to start with the error

```
Error using matlabpath
Modifying the search path is not supported by MATLAB Compiler. Remove functions that modify the search path from your MATLAB code. To make files visible to your deployed application, add the parent folder to your MATLAB session.
```

**Comment by ghostface** [comment]
Ok I did some reading but I am unable to test this as I can't compile without a license.

Changing the addpath to 
```
if ~isdeployed
    addpath(executableDir)
end
```
Should avoid the error.
I think matlab automatically includes the current directory in the searchpath so we might be able to live without this call anyway.

Can you please do a recompile with this change for testing please @bw1129 ?

**Comment by bw1129** [comment]
Unfortunately there’s bigger issues than that. Linux is a no go at the moment.

**Comment by ghostface** [comment]
It was working fine from what I could see when I was testing the non-packaged version with the evaluation license.

Is there a list or any more info on what those issues are? 
I am happy to try to help

**Comment by bw1129** [comment]
I compiled it from a version of Matlab for Ubuntu on virtual machine. The problem I had was I could not test the app there directly so I was never sure what it was doing. I’ve no experience in linux so I’m really no help. The code is there. All I can say is grab it and try to compile yourself. Sorry

**Comment by meoso** [comment]
progress :)
latest version `PIDtoolbox_v0.392` runs on Debian :)
but also i get this
![image](https://user-images.githubusercontent.com/1337447/97730853-65ceae00-1aa2-11eb-86f1-776762174c8d.png)
which is likely a separate/external issue.

**Comment by bw1129** [comment]
That’s an odd error to get because it speaks to an issue with one of Matlab’s built in functions, but only as it pertains to your operating system. It’s the first time I’ve seen it and Ptb is currently working on multiple versions of win, Mac, And Linux

**Comment by JoeriCapens** [comment]
I've been using 0.392 on my Debian machine for a while and the new 0.4x version seems awesome but there's no Linux release this time? I'm also confused why there are no recently changed files in the Git source repository?

**Comment by bw1129** [comment]
Unfortunately support for linux is just too much work for me for an open source project and I no longer have a working version of Matlab in linux. 0.392 is still there for Linux.

**Comment by bw1129** [comment]
I managed to get a draft of upcoming PTB v0.44 compiled for Linux. Should be ready for release soon.

**Comment by mituritsyn** [comment]
@bw1129 will be there linux binaries in latest releases?
